### ISSUE TITLE
Environment proposal

### ISSUE THREAD

**Issue by danbucholtz** [ionic-team/ionic-app-scripts#762]
I am thinking we could do this:

1. Introduce a new flag that gets passed in like this:

`ionic serve --env qa` or `ionic run android --env prod`

My first thought was that it's value would default to `dev` for non-prod builds, and `prod` for prod builds. Developer's can pass in whatever they want, though.

2. Based on the value that is passed in (or inferred), the `config` section of the package.json is read. It could look something like this:

```
...

"config": {
   "ionic_env" : {
       "dev" : {
          "keyToReplace" : "devValueToReplaceWith",
          "keyTwoToReplace" : "devValueTwoToReplaceWith"
       },
       "qa" : {
          "keyToReplace" : "qaValueToReplaceWith",
          "keyTwoToReplace" : "qaValueTwoToReplaceWith"
       }
   }
},
...
```

If the `ionic_env` data is not there, we would just move on in the build process. If it is present, we would then perform the text replacement.

3. Any computed values, sync or async, could be replaced after the build is done. We could either leave this up to the user to add on to the `npm script` section, or we could provide a hook into the `postprocess` step. I prefer the latter as it's easier to document and and we can probably make it a 1/2 second faster or so if we do it in app-scripts.

Feedback is appreciated.

Thanks,
Dan

**Comment by fiznool** [comment]
Looks interesting! I have a few questions. 

> Any computed values, sync or async, could be replaced after the build is done.

Could you expand on this? I didn't quite understand what you mean.

How would the app use these variables, what would the code look like? 

One of the comments in another thread mentioned keeping sensitive values out of source control. Could there be a way of reading in a value from a `process.env` variable to satisfy this requirement?

**Comment by rolandjitsu** [comment]
@danbucholtz this looks great. Angular CLI has a similar implementation.

They have a `--target` which you can use to tell the CLI to either do a production build (AOT, minification, gzip, etc.) or a dev build. Then there's the `--environment` flag which we can use to choose the appropriate environment.

I'd disagree with the `ionic_env` working as proposed though. As you also pointed out, there might be a case when we need to compute some values, and that'd be difficult for some users to figure out. How about the following `ionic_env`:
```json
"ionic_env": {
    "prod": "config/env.prod.js",
    "staging": "config/env.staging.js",
    "dev": "config/env.dev.js"
}
```

And each file would export the following:
```js
const somePackage = require('somePackage');

module.exports = {
    SOME_VAR: somePackage.getSomeValueSync(),
    SOME_ASYNC_VAL: async function getSomeValueAsync() => {
        await asyncValue = somePackage.computeSomeValueAsync();
        return asyncValue;
    }
};
```

Then `@ionic/app-scripts` would require the appropriate env based on the config from `ionic_env` and for each key it will:
* check if the value is primitive and use that if so
* or if it's a function it would invoke and
    * check if the return value is primitive and use it if so
    * if it's a Promise or an Observable, wait until it's resolved or emits and return that
* then after it successfully computed each value, it would replace where necessary

Perhaps there could be a better approach, but I thing this would work in most scenarios.

**Comment by fiznool** [comment]
Perhaps I'm missing something, but what sort of use cases would benefit from async computation of values?

**Comment by rolandjitsu** [comment]
Reading a file for instance (that's also possible to do sync). Or perhaps you store some keys in an AWS bucket and you'd like to pull those keys when you compile the app. I could think of plenty of use cases.

**Comment by fiznool** [comment]
Right, I'm with you now. :+1: 

One approach I've seen in other places is to export a function instead of an object inside the configuration file. This could simplify the logic from the app-scripts end and allow you to build the config object dynamically, and only return once you are ready.

Something along these lines:

``` js
// sync example
const somePackage = require('somePackage');

module.exports = function() {
  const config = {};

  config.SOME_VAR = somePackage.getSomeValueSync();
  config.VALUE_FROM_PROCESS = process.env.SOME_ENV_VAR_VAL;
  
  return config;
};

``` js
// async example
const somePackage = require('somePackage');

module.exports = function() {
  const config = {};

  config.SOME_VAR = somePackage.getSomeValueSync();
  config.VALUE_FROM_PROCESS = process.env.SOME_ENV_VAR_VAL;
  
  return somePackage.computeSomeValueAsync().then(val => {
    config.SOME_ASYNC_VAL = val;
    return config;
  });
};
```

app-scripts would then call this function and observe the return value:

- If the value is not a promise, assume the config object is returned synchronously.
- If the value is a promise, wait for it to be resolved and use the resolved value as the config object.

This has the benefit that app-scripts doesn't need to enumerate all the keys in the config object (what about nested keys?) and provides the most flexibility in your config script (what if the fetching of a second async value depends upon another one being resolved first?)

**Comment by rolandjitsu** [comment]
@fiznool true, that'd be much simpler.

**Comment by rolandjitsu** [comment]
Though I'd actually argue that if we export an object, app scripts could potentially run async tasks in parallel so if we'd have more than one property that evaluates to a promise/observable, it could run all of them at once and complete when all are done (something like a `Promise.all([ ... ])`/`Observable.merge(...)`).

Yet, the user could do this as well, but if app scripts would do it, it would spare the user of adding extra implementation.

I guess there are many ways this can be tackled, but the simpler the better.

**Comment by fiznool** [comment]
The difficulty there though, is that if one async value depended on
another, you wouldn't be able to resolve the configuration correctly if
app-scripts ran everything in parallel.

Better to put the logic in the hands of the user, IMHO.

**Comment by ehorodyski** [comment]
I think it could be as simple as what @rolandjitsu is proposing, however, just import the file based on what is in `process.env.ENVIRONMENT_CONSTANT` path. I think keeping it in `package.json` makes it very simple to use.

I'd also recommend trying to create an alias. I'm not sure how they are done in libraries like `@angular`, where you can use `import { HTTP} from '@angular/http'` instead of `import { HTTP } from 'angular/wherever/the/HTTP/file/lives`. This makes it easy for the user to import in their code.

**Comment by danbucholtz** [comment]
If we were to expose a hook and allow the user to load in async values, perhaps we could do something like this:

```
"config" : {
   "ionic_env" : {
      "dev" : "./scripts/dev.config.js",
      "qa" : "./scripts/qa.config.js",
      "prod" : "./scripts/prod.config.js"
      ... etc ...
   }
}
```

Those modules could then export a function that returns a `Promise`. That way it will always be async.

It could work like this:

```
module.exports = function(ionicEnvironment: string) {
   return readSomeAsyncValue().then((result: string) => {
      return {
         'keyOneToReplace' : 'valueOne',
         'keyTwoToReplace' : result
      }
   });
}
```

Something like this would work in the vast majority of use cases.

In an application's code, let's say you need to hit a different HTTP service for development and production.

You could write a service like this:

```
export class MyService {
   construtor(public http: Http){
   }

   makeServiceCall() {
      return this.http.makeRequest('$BACKEND_SERVICE_URL');
   }
}
```

Your implementation to replace it could look like this:
```
module.exports = function() {
   return Promise.resolve({
      '$BACKEND_SERVICE_URL' : 'localhost:8080/myService'
   });
}
```

**Comment by ehorodyski** [comment]
@danbucholtz - I think that is really over-complicated. 

Take a look at Webpack's [resolve aliasing](https://webpack.js.org/configuration/resolve/#resolve-alias). This could be done dynamically when starting `ionic serve` or the other scripts, depending on the `env` variable:

> `alias: {`
>      `Utilities: path.resolve(__dirname, 'src/utilities/'),`
>      `Templates: path.resolve(__dirname, 'src/templates/')`
>    `}`
>Now, instead of using relative paths when importing like so:
>
> `import Utility from '../../utilities/utility';`
>
>you can use the alias:
>
> `import Utility from 'Utilities/utility'; `

The only implication to the user is that they'd need to store their config files outside the folder with their source code, as to not include all the files defined. Or maybe not, you can run a glob that excludes the other environment files not being used. Plus, there's no resolution during application running.

**Comment by josh-m-sharpe** [comment]
For us this isn't necessarily just about just dev/staging/prod.  For some context, we're coming to ionic(v2) from a vanilla cordova app and we use the same codebase to generate multiple different apps primarily distinguished by an 'org_id'.  (The app passes this value in server calls to receive different content, amongst a bunch of other differences).  We also currently inject the internal and external version numbers since they vary (inconsistently) by org_id.

So, in a nutshell, just passing in "environment" wouldn't solve this problem for us. Why not make this feature align better with unix best practices and simply make all of the current env available inside the application somehow?

for example:

ORG_ID=123 ENVIRONMENT=staging ionic server


For what it's worth, cordova (or node?) is doing this already automatically.  That is, you can access 'process' inside a cordova hook without having to require anything.  From that, it's just process.env.ORG_ID == '123'

**Comment by rolandjitsu** [comment]
@danbucholtz I agree, that'd be a reasonable solution. And I'm sure that it would cover most use cases.

@ehorodyski I think it's already possible to use aliasing if you use your own webpack config file (I cannot confirm it, never tried it). But I do see what you mean, though that would not work if you have async operations that need to run in order to compute the env values you want to have available in the app. It would also not work for the use case @josh-m-sharpe describes where you need some env vars that are not exposed by app scripts.

@josh-m-sharpe using the solution proposed would also solve what you're referring to. You'd just need to expose from `process.env`:
```js
module.exports = function() {
    // process.env is an object and has all the env vars
    return Promise.resolve(process.env);
}
```

I think it's essential to have a more generic and universal solution, such as the one proposed by @danbucholtz in https://github.com/driftyco/ionic-app-scripts/issues/762#issuecomment-280666058, that can cover most of the uses cases since the dev needs vary as we can see from the posts on this topic.

**Comment by biesbjerg** [comment]
Am I the only one that really wants hinting of config keys, e.g. by defining a ConfigInterface? I have not seen it mentioned and I'm not sure how to do it elegantly, though...

**Comment by ehorodyski** [comment]
@biesbjerg - Right, this is what I'd like accomplished as well. I need to brush up on how to do Typescript module aliasing, so that the paths can be dynamically generated when you run `ionic serve`

**Comment by coldAlphaMan** [comment]
@danbucholtz all of the other stuff people are mentioning are great nice to haves.  That said, i think your first comment is awesome enough and will suit most needs.

**Comment by fiznool** [comment]
@danbucholtz 

> You could write a service like this:

``` ts
export class MyService {
   construtor(public http: Http){
   }

   makeServiceCall() {
      return this.http.makeRequest('$BACKEND_SERVICE_URL');
   }
}
```

> Your implementation to replace it could look like this:

``` ts
module.exports = function() {
   return Promise.resolve({
      '$BACKEND_SERVICE_URL' : 'localhost:8080/myService'
   });
}
```

Are you suggesting that `'$BACKEND_SERVICE_URL'` becomes a sort of 'magic string' that is replaced during compile time by the contents defined in the environment config file?

If so, I think this might be a bit confusing and difficult to maintain.

As others have suggested, an ideal solution would be something along the lines of the following, which could be achieved in webpack using [resolve aliasing](https://webpack.github.io/docs/configuration.html#resolve-alias):

``` ts
import { BACKEND_SERVICE_URL } from 'config';

export class MyService {
   construtor(public http: Http){
   }

   makeServiceCall() {
      return this.http.makeRequest(BACKEND_SERVICE_URL);
   }
}
```

I'm not sure how the above could be achieved with rollup. If this is a concern, a slightly lesser (but still very workable) version would be to add all config parameters to a global variable, e.g. `ENV`. With webpack, this could be achieved using the [define plugin](https://webpack.github.io/docs/list-of-plugins.html#defineplugin), and with rollup you could use the [rollup replace plugin](https://github.com/rollup/rollup-plugin-replace).

``` ts
define const ENV;

export class MyService {
   construtor(public http: Http){
   }

   makeServiceCall() {
      return this.http.makeRequest(ENV.BACKEND_SERVICE_URL);
   }
}
```

For those wishing for TypeScript autocomplete, `ENV` could be defined in `src/declarations.d.ts` as follows:

```
declare namespace ENV {
  const BACKEND_SERVER_URL: string;
  // Any other properties go here
}
```

**Comment by danbucholtz** [comment]
The resolve aliasing is interesting and wouldn't really require Webpack or Rollup to do it. From our perspective, a bundler is just a bundler, not a build tool. We don't dive into the world of plugins unless absolutely necessary, because they tighten our coupling to the tool. In our ideal world, we're going to swap out Webpack for a better bundler someday without 90% of our developers even noticing.

I think the aliasing leaves many use-cases unfulfilled where you need to load asynchronous data prior to a build. I guess you could do it ahead of time in a separate workflow/process and put it in an environment variable? I'm not sure, this seems a bit complicated to me.

I'll chat with some other team members this week and we'll figure out what we want to do. Any of these solutions will cover the vast majority of use cases.

Thanks,
Dan

**Comment by Ross-Rawlins** [comment]
+1

**Comment by ehorodyski** [comment]
>**I think the aliasing leaves many use-cases unfulfilled where you need to load asynchronous data prior to a build.** I guess you could do it ahead of time in a separate workflow/process and put it in an environment variable? I'm not sure, this seems a bit complicated to me.

Could you provide some use-cases for where you'd load async data prior to a build -- that would be environmental configurations? 

Maybe there's a disconnect in what certain people are talking about. In my view, and many others, this is analogous to having `.properties` files. They'd stay in the repo and only the specific file would get included in the build. I think making it a bunch of async resolutions just complicates matters and adds overhead to your app if you need to resolve every API URL, etc.

**Comment by riltsken** [comment]
Just my perspective as a user, but I pretty much use environment variables for any configuration changes required among different environments or deploys. This means I can change a configuration value for prod, staging, dev, individual branch builds ... any `N+ environments`. Supporting N+ environments is easy with environment variables. Just use one template. When talking about property files though it becomes harder because now you need a property file per environment or to generate a property file per environment.

I'll post my specific project use cases that we have currently:

```
maxCacheAgeInDays
ionicCloudChannel
apiUrl
oauthClientId
logLevel
logentriesToken
googleMapsJavascriptApiUrl
```

In the individual threads I see this asked more than once for environment variable support. I could generate a property file I guess at buildtime, but that would basically be me duplicating what I am today in a different form. Below are some references to what I would consider are analogous solutions for the same problem.

references:
 * https://github.com/driftyco/ionic-app-scripts/issues/370#issuecomment-271614462
 * https://github.com/driftyco/ionic-app-scripts/issues/760#issuecomment-280380299

Anyway just some thoughts. I would really love environment variable support :).

**Comment by vovikdrg** [comment]
Angular-cli has implemented and its easy to configure it in angular-cli file why not to do the same?

https://github.com/angular/angular-cli/blob/3c3f74c060cc79f3872230a18eb0c1dd8371065f/packages/%40angular/cli/blueprints/ng2/files/angular-cli.json#L24

**Comment by nottinhill** [comment]
We want this too!

**Comment by juarezpaf** [comment]
We're currently using a script that runs before `ionic serve` and generate the correct file for us.
Let me explain it in more details here:

1. We created an `env` folder in the root of the Ionic project and added this folder to our `.gitignore`
2. Inside of the `env` folder we have different files:
```
./env
  production.json
  sandbox.json
  staging.json
```
3. Each file has the variables we need, for example, our `staging.json` file looks like:
```
{
  "api_url": "'https://staging.api.url'",
  "client_id_login": "'566d41f...'",
  "client_secret_login": "'511480878...'",
  "google_maps_key": "'tEkoJK...'"
}
```
4. We changed the `scripts` section inside of the `package.json` file to have a `preionic` phase and also an independent command:
```
"scripts": {
    "ionic:build": "ionic-app-scripts build",
    "ionic:serve": "ionic-app-scripts serve",
    "preionic:build": "node ./scripts/replace.env",
    "preionic:serve": "node ./scripts/replace.env",
    "generate-env": "node ./scripts/replace.env",
    "test": "ng test"
}
```
5. Our `replace.env.js` file handles the generations of our `config.ts` file https://gist.github.com/juarezpaf/7a9007dfcef25ec24f7de900df8fe04e
6. The `config.ts.sample` is very simple and it'll just expose these variables to use use across our app
```
export class AppConfig {
  static get API() {
    return ENV.api_url;
  }

  static get GOOGLE_MAPS_KEY() {
    return ENV.google_maps_key;
  }

  static get CREDENTIALS() {
    const credentials = {
      'login': {
        'client_id': ENV.login.client_id,
        'client_secret': ENV.login.client_secret
      }
    }
    return credentials;
  }
}
```
7. We use the `AppConfig` in our AuthService:
```
import { AppConfig } from '../config/config';

@Injectable()
export class Auth {
  credentials: any;
  private apiUrl = AppConfig.API;
  
  constructor(public http: Http) {
    credentials.client_id = AppConfig.CREDENTIALS.login.client_id;
    credentials.client_secret = AppConfig.CREDENTIALS.login.client_secret;
  }
}
```
8. With everything in place we can use it in 3 different ways:
  * `npm run generate-env --sandbox`
  * `ionic serve --staging`
  * `ionic build:ios --production`

I found some *cons* with this `replace.env.js` approach:
 * Local files with tokens and other sensitive info
 * If we need other environments we need to change this file;
 * Write the variables into a the `config.ts` file every time;

**Comment by jbrownssf** [comment]
Hey guys, I have been working all day on a temporary solution to this thread that should work for most of the use cases I have read in this and a few other threads of the same topic. The main concerns I noticed form suggestions include:
1. Fear of an update that could break a solution
2. Unable to use environment variables
3. Required changing the text in a file to switch environments
4. Required cardinal knowledge of the script and how it functions (personally)

My project should solve all of those issues. If there are any others I am happy to work on it.
[Git Repository](https://github.com/jbrownssf/ionic-environment-flags)

I have it fairly well spelled out in the README.md.

**Comment by pbowyer** [comment]
There is a very interesting implementation at https://github.com/driftyco/ionic-app-scripts/pull/683#issuecomment-287401855 which I'm using successfully.

**Comment by vkniazeu** [comment]
@pbowyer , does it work for you with `ionic build browser --prod`?
I've commented on #683 as it fails for me when the AoT Compiler kicks in.

**Comment by vkniazeu** [comment]
Solution in #683 fails with
`[17:52:03]  Error: Error encountered resolving symbol values statically. Could not resolve @app/config relative to XXX/src/app/app.module.ts., resolving symbol AppModule in...`

**Comment by pbowyer** [comment]
@vkniazeu That exact command works for me - but I can't see anything in the output saying AoT is running, so can't definitively confirm.

**Comment by vkniazeu** [comment]
@pbowyer , thanks! It's the `ngc started ...` line which indicates that angular-cli AoT compiler is being used. It should be default if the `--prod` flag is used.
Are you using latest Ionic and Angular packages in your project?

**Comment by pbowyer** [comment]
@vkniazeu That line's present in the output, so it's being used. It's not present if I omit `--prod`.

As to versions, I'm on:
```
Cordova CLI: 6.5.0
Ionic Framework Version: 2.2.0
Ionic CLI Version: 2.2.1
Ionic App Lib Version: 2.2.0
Ionic App Scripts Version: 1.1.4
ios-deploy version: 1.9.1
ios-sim version: 5.0.13
OS: OS X El Capitan
Node Version: v6.10.0
Xcode version: Xcode 8.2.1 Build version 8C1002
```

**Comment by vkniazeu** [comment]
@pbowyer , thank you for your time. Strangely, it fails for another person as well with the same error and same setup. My package versions are newer, but I don't think this is the reason for the ` resolving symbol values statically` error. I'll keep digging.

**Comment by rolandjitsu** [comment]
@pbowyer The solution described in https://github.com/driftyco/ionic-app-scripts/pull/683 does not work with AOT. And `ngc` should fail for `import ... from '@app/config'` as it will not be able to resolve it since it's never declared.

**Comment by vkniazeu** [comment]
@rolandjitsu , thanks for bringing this again. Do you have an idea of how to overcome this lack of declaration for AoT to work? Or, perhaps, you have an alternative solution that would be similar in simplicity?

**Comment by rolandjitsu** [comment]
@vkniazeu I currently use Rollup instead of Webpack and [rollup-plugin-replace](https://github.com/rollup/rollup-plugin-replace).

`rollup.config.js`:
```js
// For reference, check the following:
// https://github.com/driftyco/ionic-app-scripts/blob/master/config/rollup.config.js
// Config borrowed from: https://github.com/driftyco/ionic-cli/issues/1205#issuecomment-255744604
const nodeResolve = require('rollup-plugin-node-resolve');
const commonjs = require('rollup-plugin-commonjs');
const globals = require('rollup-plugin-node-globals');
const builtins = require('rollup-plugin-node-builtins');
const json = require('rollup-plugin-json');

// We use the replace plugin to provide the correct env variables
const replace = require('rollup-plugin-replace');

// We will use the machine ip to make HTTP requests to the local machine from an external device while in dev mode.
const ip = require('ip');
// We need to expose the app id to the app.
const ionicConfig = require('../ionic.config.json');


// NOTE: If IONIC_ENV is not set, we check if the CLI args contain the '--prod' flag.
function isProd() {
	return process.env.NODE_ENV === 'production' || process.env.IONIC_ENV === 'prod' || process.argv.slice(2)
			.some((arg) => arg.indexOf('--prod') !== -1)
}


// https://github.com/rollup/rollup/wiki/JavaScript-API
const config = {
	/**
	 * entry: The bundle's starting point. This file will
	 * be included, along with the minimum necessary code
	 * from its dependencies
	 */
	entry: process.env.IONIC_APP_ENTRY_POINT,

	/**
	 * sourceMap: If true, a separate sourcemap file will
	 * be created.
	 */
	sourceMap: true,

	/**
	 * format: The format of the generated bundle
	 */
	format: 'iife',

	/**
	 * dest: the output filename for the bundle in the buildDir
	 */
	dest: process.env.IONIC_OUTPUT_JS_FILE_NAME,

	/**
	 * plugins: Array of plugin objects, or a single plugin object.
	 * See https://github.com/rollup/rollup/wiki/Plugins for more info.
	 */
	plugins: [
		builtins(),
		commonjs(),
		nodeResolve({
			module: true,
			jsnext: true,
			main: true,
			browser: true,
			extensions: ['.js']
		}),
		globals(),
		json(),
		replace({
			values: {
				'{{API_HOST}}': isProd() ? 'https://api.domain.com' : `http://${ip.address()}:8888`,
				'{{IONIC_ENV}}': isProd() ? 'prod' : 'dev',
				'{{APP_ID}}': ionicConfig.app_id
			},
			// Config
			exclude: 'node_modules/**'
		})
	]
};


module.exports = config;
```

`env.ts` (I placed it under `src/app/`):
```ts
export const env: any = '{{IONIC_ENV}}';
export const apiHost = '{{API_HOST}}';
export const appId = '{{APP_ID}}';

export function isProd(): boolean {
	return env === 'prod';
}
```

And you just import from `env` wherever you need. It should work with AOT as well.
Though I'm encountering other issues with Rollup that are unrelated.

I'm not sure about a fix for `@app/config`, I imagine making a declaration file should suffice, but I'm not sure how ngc will work with it since it needs to resolve the values statically.

**Comment by vkniazeu** [comment]
@rolandjitsu , thank you for sharing your setup. I will consider it as my project grows if an Ionic own solution is not available by then.

**Comment by mumairofficial** [comment]
I got too many ideas for env variable setup and now fully confused about what is best for scale-able enterprise application? should I wait for ionic team implementation for ionic-3

**Comment by omRm** [comment]
Too many solutions are out there right now, most of them are not fully functional with prod development. 
We need a official solution and support from the ionic team for this.

**Comment by rossholdway** [comment]
Hello. Based off the great work by @juarezpaf in his previous comment, we're using something very similar.

Advantages for us are with this approach are that it's simple, does not involve changing webpack or Rollup config, and follows the original --env [environment] flag proposal (so we can hopefully switch to an official implementation when released).

We are able to use the standard `ionic serve` command, and **default to a development env**, or use the --env flag when we require a different environment e.g. `ionic build ios --prod --env production`

Our setup is as follows (very similar to juarezpaf with a few modifications!).

1. We have a .env folder containing json files for each environment (added to .gitignore)
```
./env
  production.json
  staging.json
  development.json
```

Example of development.json
```
{
  "environment": "development",
  "endpoints": {
    "server1": "https://....",
    "server2": "https://...."
  }
}
```

2. Reference the script in package.json available at https://gist.github.com/rossholdway/16724496806b66a162ee6cbf8bfc5def

```
"scripts": {
    ...
    "preionic:build": "node ./scripts/environment",
    "preionic:serve": "node ./scripts/environment",
    ...
  }
```

This will generate `app/app.config.ts`.

3. In your module import the config with `import { APP_CONFIG, AppConfig } from './app.config';` and add as a provider
```
providers: [
   ...
    {
      provide: APP_CONFIG,
      useValue: AppConfig
    }
  ]
```

4. Now you can import your environment specific app config where needed via Angular DI. e.g.


```
import { APP_CONFIG } from './app.config';

@Component({
  selector: 'example-page',
  templateUrl: 'example-page.html'
})
export class ExamplePage {
  rootPage = TabsPage;

  constructor(@Inject(APP_CONFIG) private config) {
    console.log(config.environment);
    console.log(config.endpoints.server1);
  }

}
```

There are some disadvantages, mainly that we can't use TypeScript in app.config.ts easily, but apart from that we are yet to run into any issues.

**Usage examples**
`ionic serve` (Not specifying the --env flag will default to using .env/development.json)
`ionic build ios --prod --env production` (Will use .env/production.json to build app.config.ts)

**Comment by danbucholtz** [comment]
app-scripts is a never ending project. I'll hopefully have some time to tackle this in the coming weeks.

Thank you for being patient. 💯 

Thanks,
Dan

**Comment by ekoome** [comment]
+1

**Comment by the-ult** [comment]
@danbucholtz any news on your process? Or what would be a proper solution?
Would the solution @rossholdway provided, be the best solution for now?

**Comment by danbucholtz** [comment]
Soon. Very soon! So here's what on my road map:

1. Migrate to a "fesm" build of ionic-angular. It did not make it for this weeks release, but should be in in two weeks in the next release. This is basically bundling code with rollup before bundling it again with webpack. The reason for this is to avoid unnecessary closures. Our testing shows this should knock off around ~1 second of app start-up time on mobile devices. (I estimate this will take 2 more days)

2. Make some fixes to navigation & deep linking in ionic-angular (I estimate this will take two weeks).

I think that's all I have that is higher priority than the environment stuff. We'll get this in soon.

Thanks,
Dan

**Comment by trumbitta** [comment]
I personally reckon this issue should be higher priority than most, but I'm glad you'll be finally working on it 🎉

**Comment by yuricamara** [comment]
Folks,
This solution works like a charm. I implemented it in my project.  I even think that it removes this issue from so high priority.

[Ionic 2 Environment Variables: The Best Way](http://roblouie.com/article/296/ionic-2-environment-variables-the-best-way/)

**Comment by trumbitta** [comment]
@rossholdway you solution works for me only if I use `npm run ionic:serve` instead of `ionic serve`. Any hints?

**Comment by trumbitta** [comment]
Nevermind, it's working. It just doesn't log anything when run via `ionic serve`.

**Comment by rossholdway** [comment]
@trumbitta Sorry, I'm not sure why it wouldn't be working with `ionic serve` for your setup. I'd just stick with using `npm run ionic:serve` for now as it sounds like it won't be long until we have support for environments built directly into app scripts 🎉

**Comment by trumbitta** [comment]
Hi danbucholtz, sorry to bother you and thanks for your work!  
Any news?

**Comment by RoopeHakulinen** [comment]
This would be really essential for me also. Anything that can be done to help you guys?

**Comment by rolandjitsu** [comment]
@trumbitta and @RoopeHakulinen there are plenty of solutions through some of the comments here that work well (for most cases), perhaps you could elaborate on what's missing so that the team can better understand the requirements (being a discussion and all ...)?

**Comment by RoopeHakulinen** [comment]
@rolandjitsu Because they are hackish ways to implement something that should be provided by Ionic app scripts instead. The requirements seem to really clear and I don't think that is the problem with this one. For requirements one can simply take a look at how it is achieved in Angular CLI.

**Comment by rolandjitsu** [comment]
@RoopeHakulinen I won't argue about what should and what should not be 😄 Just bear in mind that this is open source, so nobody is stopping you from forking it and making your impl. of app-scripts that provides a *non-hacky* solution for env. vars 👍

And you can either use the existing solutions provided by some of the comments above or not 😄

**Comment by trumbitta** [comment]
@rolandjitsu I'm not sure what your point is.

Mine was about kindly and respectfully asking @danbucholtz about something he said a month ago he would be start working on in a couple weeks.

From https://github.com/ionic-team/ionic-app-scripts/issues/762#issuecomment-300224592 - I emphasized the relevant bits:

> **Soon. Very soon! So here's what on my road map:**

> Migrate to a "fesm" build of ionic-angular. It did not make it for this weeks release, but should be in in two weeks in the next release. This is basically bundling code with rollup before bundling it again with webpack. The reason for this is to avoid unnecessary closures. Our testing shows this should knock off around ~1 second of app start-up time on mobile devices. **(I estimate this will take 2 more days)**

> Make some fixes to navigation & deep linking in ionic-angular **(I estimate this will take two weeks)**.

> **I think that's all I have that is higher priority than the environment stuff. We'll get this in soon.**

> Thanks,
> Dan

**Comment by rolandjitsu** [comment]
@trumbitta the point was that there are existing solutions and this is still an open issue, so most likely it's going to be implemented at some point in time 😄  Sorry if my comment was perhaps misleading (I'm a bit tired).

But let's be honest, time estimates in the dev world can always go two ways, it either takes way longer than anticipated or way less 😆

**Comment by trumbitta** [comment]
@rossholdway the script can be now run with `ionic-cli@3.4.0`: https://github.com/ionic-team/ionic-cli/blob/master/packages/cli-plugin-gulp/README.md

**Comment by biesbjerg** [comment]
@trumbitta any idea how you would determine if it's a dev or prod build using gulp?

`my-project/gulpfile.js`
```js
var gulp = require('gulp');

gulp.task('ionic:watch:before', copyEnvironmentConfig);
gulp.task('ionic:build:before', copyEnvironmentConfig);

function copyEnvironmentConfig() {
	console.log(process.env); // Nothing here to determine which config file we should copy
}
```

**Comment by trumbitta** [comment]
sorry @biesbjerg , I passed on Gulp and directly went from Grunt to NPM scripts

**Comment by trumbitta** [comment]
I ended up implementing my own solution: https://www.williamghelfi.com/blog/2017/06/22/ionic-environments-webpack/

**Comment by RoopeHakulinen** [comment]
Thank you @trumbitta! This is a solution that is both not too hacky and future-proof. Works like a charm and easily understandable. Great job!

**Comment by jcezarms** [comment]
@danbucholtz that's great news! I've been struggling with this for a while now since Ionic projects of mine are soon to enter homologation/support phase. Ended up implementing something similar to @trumbitta's solution.

Please get back at us if you reach an advance/barrier for the implementation. :weight_lifting_man:

**Comment by sebaferreras** [comment]
Thanks @trumbitta, your solution looks great. I was using a similar solution based on a webpack alias, but what I don't like about using `--prod` flag to decide which configuration details are going to be used, is that most of the times I want to run the app by using the `--prod` flag but just to **test** the app as the end user will be using it, but I don't want to use the prod analytics key, the prod api, and so on... So I guess that what @danbucholtz said at the beginning of this thread would be the best way to manage this issue:

> I am thinking we could do this:
> 
> Introduce a new flag that gets passed in like this:
> ionic serve --env qa or ionic run android --env prod
> 
> My first thought was that it's value would default to dev for non-prod builds, and prod for prod builds. Developer's can pass in whatever they want, though.

> [...]

**Comment by wursta** [comment]
@trumbitta, thank you for your solution. I have some porblem when i build my app with --release flag. ngc fails with foolowing error: 

ionic cordova build android --release --prod
Running app-scripts build: --prod --iscordovaserve --externalIpRequired --nobrowser

[18:57:16]  build prod started ...
[18:57:16]  clean started ...
[18:57:16]  clean finished in 3 ms
[18:57:16]  copy started ...
[18:57:16]  ngc started ...
[WARN] Error occurred during command execution from a CLI plugin (@ionic/cli-plugin-cordova). Your plugins may be out of
       date.
Error: Error encountered resolving symbol values statically. Reference to a local (non-exported) symbol 'webpackGlobalVars'. Consider exporting the symbol (position 3:15 in the original .ts file), resolving symbol Configuration.appConfig in D:/_PROJECTS/RNSWebMobile/mobileapp/src/providers/configuration.ts, resolving symbol AppModule in D:/_PROJECTS/RNSWebMobile/mobileapp/src/app/app.module.ts, resolving symbol AppModule in D:/_PROJECTS/RNSWebMobile/mobileapp/src/app/app.module.ts, resolving symbol AppModule in D:/_PROJECTS/RNSWebMobile/mobileapp/src/app/app.module.ts

**Comment by trumbitta** [comment]
@wursta can you provide a plunkr or something with your code so I can take a look at it? I successfully released an app with this technique.

**Comment by Hesesses** [comment]
Any news about the "official" solution?

**Comment by danbucholtz** [comment]
Sorry, I have been super busy with nav changes lately. They took longer than planned, and now my time has been allocated to a different task for the framework. Does anyone wanna do a PR for this? If so, reach out to me at `danb` on Ionic slack. I can help you get started. @rolandjitsu, any interest?

Thanks,
Dan

**Comment by rolandjitsu** [comment]
@danbucholtz I might be able to take a shot at this, but it depends on how long the impl. would take, because I also have some holiday planned soon 😄  Let's chat over slack. Or maybe someone else is interested ...

**Comment by danbucholtz** [comment]
I don't think it will be too bad to be honest, but I'm not 100% sure either. Let's chat. The offer stands for anyone interested. I will help.

Thanks,
Dan

**Comment by janein** [comment]
@wursta I've got the same issue too.
`ionic cordova run android` rund perfectly, while `ionic cordova run android --prod` fails with `Error occurred during command execution from a CLI plugin (@ionic/cli-plugin-cordova). Your plugins may be out of date.`

Strange thing is, that it worked for me last week.
Only thing I did since then was to remove and re-add android as a platform.

Did you find a solution for this?

**Comment by janein** [comment]
@trumbitta hey, I really appreciate your soluition above.
It worked until I updated ionic. It sillt works as long as I do not build in prod-mode.
In prod mode I get:

`Error: Error encountered resolving symbol values statically. Reference to a local (non-exported) symbol 'webpackGlobalVars'. Consider exporting the symbol, resolving symbol ConfigurationService in /../src/providers/configuration.ts, resolving symbol AppModule in /../src/app/app.module.ts, resolving symbol AppModule in /../src/app/app.module.ts, resolving symbol AppModule in /../src/app/app.module.ts`

Did you encounter the same problem or do you have an idea how to avoid this?
thanks! :)

**Comment by raxkaynan** [comment]
@janein I've encountered the same error yesterday. I believe the reason for it is that you're using template literals (i.e. a value wrapped in backticks).

**Comment by janein** [comment]
@raxkaynan so did you find a solution?

**Comment by raxkaynan** [comment]
@janein I thought you already had an idea of it. The solution is simply not to use template literals. Instead, just use the '+' operator for concatenation (e.g. var1 + 'some value').

**Comment by janein** [comment]
@raxkaynan oh I see there was a missunderstanding - I never use those configs in template literals. That's what got me confused when i read your comment ;)
I'm still struggling to get this working - seems like this is a pretty hot topic for ionic-devs right now. Especially my use-case where these configs are required to use different firebase-api-keys inside the `app.module.ts`.

**Comment by pbowyer** [comment]
This is my take on multiple environments: https://github.com/pbowyer/ionic-cascading-config

I wanted to be able to extend configurations which was the driver behind this approach. Autocomplete works in my IDE when typing environment variables.

Currently it's working for `ionic serve` but not for `ionic cordova build` as I can't work out how to pass the environment variable through (command-line arguments aren't passed). I'm looking at @roblouie's approach in http://roblouie.com/article/296/ionic-2-environment-variables-the-best-way/ for inspiration using `NODE_ENV`, but so far `npm run build` fails every time (is it meant to be equivalent to `ionic cordova build ios`?)

Feedback very welcome. There's a lot of different solutions in this thread, every one with their own take on how config files should be formatted, merged, stored and delivered.

**Comment by okonon** [comment]
@danbucholtz hello, i was wondering if this feature is on the roadmap for app scripts? This feature will make a lot of people's lives easier!

**Comment by luckylooke** [comment]
@trumbitta solution is not enough for me, I need to import env.dev.ts or env.prod.ts according to IONIC_ENV, I have been using webpack resolve.alias in older versions of Ionic. Now aliases dont work. :/

Why do I need ts files? Because inside env.dev.ts I can import other modules needed only in dev environment, so they are not bundled in prod version.

**Comment by pbowyer** [comment]
@luckylooke Give my version a go which uses .ts files. It's not using IONIC_ENV but you can easily modify it to.

- https://github.com/pbowyer/ionic-cascading-config

**Comment by starstuffharvestingstarlight** [comment]
My favourite solution so far is @pbowyer 's. I'm using a similar solution but using Environment objects. Here's an example gist: https://gist.github.com/starstuffharvestingstarlight/b41ff3cc4aeacecef55e0bcb6981e78f

Note that this would solve most problems referred in these threads, including validation of environment values, which might be good for big teams.

In general, I think it'd be ideal to make no assumptions about what the environment object/constant is, and instead just replace the file being loaded. Depending on the use case, this can be a simple hash of strings, or many complex configuration strings/generator functions/etc.

**Comment by luckylooke** [comment]
I have found the way to make aliases work again, so now I am using alias to switch between env files, which is perfect solution for me ;)
Credit: https://github.com/ionic-team/ionic-app-scripts/pull/683#issuecomment-294078919

**Comment by agasbzj** [comment]
Like @starstuffharvestingstarlight said,  hope to use angular-cli's solution in ionic too.

**Comment by danbucholtz** [comment]
We are going to do this for 3.1.0 in two weeks.

Thanks,
Dan

**Comment by danbucholtz** [comment]
I am not sure exactly how this will look yet as I try to create a strawman, so I will hack a bit and ask for feedback.

Thanks,
Dan

**Comment by seiyria** [comment]
@danbucholtz if it helps, the only approach I've had success with so far is @pbowyer's cascading config. Although, I've had mixed success with it - the environment default has to be ts, and the env files need to be js. All of them TS will complain about something, and all of them JS will complain about overwriting the source file. Basically, if you can do what he has there, but more elegant, that would be ideal, I think.

**Comment by gshigeto** [comment]
@danbucholtz I also put together another solution that has been working great for me and others that I posted about [here](https://github.com/ionic-team/ionic-cli/issues/1205#issuecomment-326344949).
I also set up a sample repo [here](https://github.com/gshigeto/ionic-environment-variables).

I created an alias that references the `IONIC_ENV` which points to an environment file in my `src` directory.

This way it is
* Easy to use `import {ENV} from @app/env`
* You can even use it in your `app.module.ts`

**Comment by Sneko** [comment]
@danbucholtz any news when it will be released? Thanks

**Comment by Sneko** [comment]
@danbucholtz are you still working on this? Thanks :)

**Comment by mhartington** [comment]
Hey folks, just chiming in here. @danbucholtz is out for a bit (new baby 🎉 ) 

We'll pick this up in a bit.

**Comment by okonon** [comment]
Nice!

**Comment by capcom-r** [comment]
Looking very forward to this also, wishing you the best @danbucholtz

**Comment by agasbzj** [comment]
I wonder since 3.1.0 is out, is this proposal accomplished?

**Comment by murraybauer** [comment]
Don't believe so - see the ionic [change log](https://github.com/ionic-team/ionic/blob/master/CHANGELOG.md)

**Comment by yuricamara** [comment]
Why "see the ionic change log", @murraybauer ? I found nothing about env there.

**Comment by mlegenhausen** [comment]
If someone still looks for the `environments` file way like in the angular-cli. You can use this custom `webpack.config.js` file:

```js
const path = require('path');

const webpack = require('webpack');
const config = require('@ionic/app-scripts/config/webpack.config.js');

config.prod.plugins.push(new webpack.NormalModuleReplacementPlugin(
  /src\/environments\/environment\.ts/,
  path.resolve('./src/environments/environment.prod.ts')
));

module.exports = config;
```

Don't forget to add `"config": { "ionic_webpack": "./webpack.config.js" }` to your `package.json` and to create the appropriate files under `src/environments`.

**Comment by AksimO** [comment]
Take a look at plugin [webpack-environment-suffix-plugin](https://github.com/AksimO/webpack-environment-suffix-plugin), similar approach to what @mlegenhausen suggests.
```js
const EnvironmentSuffixPlugin = require('webpack-environment-suffix-plugin');
//...
new EnvironmentSuffixPlugin(
    ext: 'ts'
    suffix: process.env.NODE_ENV
);
```

**Comment by agasbzj** [comment]
@AksimO This works perfectly, THANK YOU!!

**Comment by alexbonhomme** [comment]
@mlegenhausen Your solution is exactly what I was looking for ! I don't understand why the Ionic team doesn't include this feature :/

**Comment by emcniece** [comment]
https://github.com/ionic-team/ionic-app-scripts/issues/762#issuecomment-344223929 is AOT-compatible, which is a crucial feature for me. I am including environment-specific variables in the main `@NgModule` declaration:

```
@NgModule({
    imports: [
        CommonModule,
        BrowserModule,
        HttpModule,
        PagesModule,
        ComponentsModule,
        StoreModule.provideStore(reducer),
        EffectsModule.run(DeviceEffects),
        IonicModule.forRoot(AppComponent,
            // Disable page transitions for screenshot task
            {animate: environment.enableAnimation}
        ),
       ....
```

This can only be done by including a non-NgModule component - `ngc` fails if you use a solution like https://github.com/ionic-team/ionic-app-scripts/issues/762#issuecomment-285053251.

**Comment by Daskus1** [comment]
Are we going to see this feature any time soon?

**Comment by luckylooke** [comment]
Angular CLI has parfect solution, which best fits my needs, [see here](https://github.com/angular/angular-cli/wiki/stories-application-environments). It would be very nice to have same mechanism in ionic projects. Thanks ;)

**Comment by danbucholtz** [comment]
We put this feature on hold because we are moving to the Angular CLI 🎉 .

Thanks,
Dan

**Comment by keithdmoore** [comment]
@danbucholtz Any idea as to when the ionic-cli will be replaced by the Angular-CLI?

**Comment by danbucholtz** [comment]
As soon as possible. No date yet but it will coincide with Ionic 4.

Check out the `core` branch of the Ionic repo if you're interested in following along.

Thanks,
Dan

**Comment by keithdmoore** [comment]
Thanks. That helps.  I will need to implement something prior to that.

**Comment by emcniece** [comment]
@keithdmoore:

1. Extend `webpack.config.js`, inject your own `customEnvPlugin`:

```js
// Extends ./webpack.default.config.js
// which is a direct copy of node_modules/@ionic/app-scripts/config/webpack.config.js
//
// We force devs to copy/paste to webpack.default.config.js so that upgrading
// @ionic/app-scripts is explicit, intentional, and reviewed.
//
// Added minimist, customEnvPlugin

var defaultConfig = require('./webpack.default.config.js');
var webpack = require('webpack');
var path = require('path');
var fs = require('fs');
var chalk = require('chalk');

var argv = require('minimist')(process.argv.slice(2));

// Build environment vars
var customEnvPlugin;

if(!!argv.env){
  fileName = './src/environments/environment.'+argv.env+'.ts';
  var fileExists = fs.existsSync(fileName);

  if(fileExists){
    console.log(chalk.bgGreen('CONFIG LOADED: ', fileName));

    customEnvPlugin = new webpack.NormalModuleReplacementPlugin(
      /src\/environments\/environment\.ts/,
      path.resolve('./src/environments/environment.'+argv.env+'.ts')
    );
  } else {
    console.log(chalk.bgRed('CONFIG SPECIFIED BUT MISSING: ', fileName));
    process.exit();
  }
} else {
  console.log('CONFIG DEFAULT: ', './src/environments/environment.ts');
  customEnvPlugin = new webpack.NormalModuleReplacementPlugin(
    /src\/environments\/environment\.ts/,
    path.resolve('./src/environments/environment.ts')
  );
}

defaultConfig.prod.plugins.push(customEnvPlugin);
defaultConfig.dev.plugins.push(customEnvPlugin);

module.exports = defaultConfig;
```

2. Create `src/environments/environment.ts`, place your variables:

```js
// Dev profile: animations & devtools

export const environment = {
  production: false,
  enableAnimation: true,
  enableDevTools: true,
  settings: {}
};
```

2.5 Clone `src/environments/environment.ts` to `src/environments/environment.dev.ts` or `src/environments/environment.prod.ts` as needed. The `.dev` and `.prod` are used in CLI

3. Import into modules as needed. Example: `src/app/app.module.ts`:

```js
...
import { environment } from '../environments/environment';
...

@NgModule({
    imports: [
        CommonModule,
        BrowserModule,
        HttpModule,
        PagesModule,
        ComponentsModule,
        ...
        IonicModule.forRoot(AppComponent,
            // Disable page transitions for screenshot task
            {animate: environment.enableAnimation}
...
```

4. Run via CLI with flag:  `ionic serve -b --env=dev`

**Comment by mlegenhausen** [comment]
I have to step back from my solution all these webpack solutions fail when you run the production build (`--prod`). The problem is that `ngc` is executed before the webpack configuration does happen. That means that during the static type analyses your default `environment.ts` is used and statically compiled in your application. After that `webpack` is executed and your `environment.ts` is replaced by e. g. `environment.prod.ts`. Now do you have following problem. When you try `console.log(environment)` you see your production configuration, but because ngc deconstructs your environment object during the compilation step everywhere where have used an `environment` variable in your `@NgModule` configuration it stayes the default configuration. This error is hard to track down so I would not recommend to replace any module via webpack.

My recommendation is to you the angular cli with ionic when you need environment file support!

**Comment by emcniece** [comment]
@mlegenhausen make the prod config the default environment file? Then when not using `ngc` you can specify `--env=dev`and let webpack rewrite without issues. 

This is working well for me so far.

**Comment by mlegenhausen** [comment]
This does only work when you have only one production version. I need to configuer differntly for android ios and web.

**Comment by ramonornela** [comment]
@danbucholtz can apply a solution similar to https://github.com/facebookincubator/create-react-app/blob/master/packages/react-scripts/template/README.md#adding-custom-environment-variables, following https://12factor.net/config see too: https://www.npmjs.com/package/dotenv.

**Comment by GFoley83** [comment]
@emcniece Does you solution support using a custom environment file e.g. `environment.test.ts` when using `ngc`? If not what modification would be needed?
I'd like to be able to build APKs for the various environments e.g. `ionic cordova build android --test`.

**Comment by emcniece** [comment]
@GFoley83 Yes, AOT (`ngc`) compilation works well with this pattern.

The downside of the provided example is that "production" builds end up with double flags, but that's not the end of the world. Here's a sample of my `package.json` scripts:

```
"build-aot": "ionic-app-scripts build --release --prod --env=prod",
"production-ios": "ionic cordova build ios --buildConfig ./config/build-ios.json --release --prod --env=prod",
"production-android": "ionic cordova build android --release --prod --env=prod",
"production-browser": "ionic cordova build browser --release --prod --env=prod",
```

The `--prod` flag is for Ionic or Cordova, not totally sure which at this moment. The `--env=prod` is the custom flag for the provided example Webpack config modification. Using `--prod` compiles the code with `ngc`.

**Comment by murraybauer** [comment]
I am just awaiting ionic 4, which will enable angular CLI support, then
finally there will be an OTB solution.

**Comment by GFoley83** [comment]
I ended up using a combo of @emcniece and @gshigeto solutions, so massive thanks to them.

I wanted to have environment variables work in Ionic exactly like how they do in Angular CLI (especially given Ionic 4 is somewhere on the horizon), whilst also supporting Karma and Protractor tests. 

To use the environment variables, just add `import { environment } from '@env/environment';`.
Also supports `ngc` so I can build an Android APK with any environment config e.g. my `environment.test.ts` config.

I've managed that with the following setup:

**/src/app/environments** folder contains 
```
environment.model.ts
environment.ts
environment.prod.ts
environment.test.ts
```

**/tsconfig.json**

```json
{
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "baseUrl": "./src",
    "declaration": false,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "paths": {   
      "@env/*": [
        "environments/*"
      ]
    },
    "lib": [
      "dom",
      "es2016"
    ],
    "module": "es2015",
    "moduleResolution": "node",
    "sourceMap": true,
    "target": "es5"
  },
  "include": [
    "src/**/*.ts"
  ],
  "exclude": [
    "node_modules",
    "src/**/*.spec.ts"
  ],
  "typeRoots": [
    "node_modules/@types"
  ],
  "compileOnSave": false,
  "atom": {
    "rewriteTsconfig": false
  }
}
```

**/config/webpack.config.js**

```js
var chalk = require("chalk");
var defaultConfig = require('@ionic/app-scripts/config/webpack.config.js');
var ionic_env = process.env.IONIC_ENV;
var fs = require('fs');
var path = require('path');
var env = require('minimist')(process.argv.slice(2)).env || process.env.IONIC_ENV || 'dev';
var webpack = require('webpack');

defaultConfig.dev.resolve.alias = {
    "@env/environment": path.resolve(environmentPath('dev'))
};

defaultConfig.prod.resolve.alias = {
    "@env/environment": path.resolve(environmentPath('prod'))
};

console.log(chalk.yellow.bgBlack('\nUsing ' + env + ' environment variables.\n'));

if (!!env) {
    var pathToCustomEnvFile = path.resolve(environmentPath(env));

    defaultConfig[env] = defaultConfig[ionic_env];
    defaultConfig[env].resolve.alias = {
        "@env/environment": pathToCustomEnvFile
    };

    var customEnvPlugin = new webpack.NormalModuleReplacementPlugin(
        /src\/environments\/environment\.ts/, pathToCustomEnvFile
    );

    defaultConfig.prod.plugins.push(customEnvPlugin);
    defaultConfig.dev.plugins.push(customEnvPlugin);
}

module.exports = function() {
    return defaultConfig;
};


function environmentPath(env) {
    var filePath = './src/environments/environment' + (env === '' ? '' : '.' + env) + '.ts';
    if (fs.existsSync(filePath)) {
        return filePath;
    }

    console.log(chalk.red.bgWhite('\n' + filePath + ' does not exist!\n'));
    process.exit();
}
```

**/test-config/webpack.test.js**

```js
var webpack = require('webpack');
var path = require('path');
var fs = require('fs');
var chalk = require('chalk');

module.exports = {
    devtool: 'inline-source-map',

    resolve: {
        alias: {
            '@env/environment': path.resolve(environmentPath('dev')),
        },
        extensions: ['.ts', '.js']
    },

    module: {
        rules: [{
                test: /\.ts$/,
                loaders: [{
                    loader: 'ts-loader'
                }, 'angular2-template-loader']
            },
            {
                test: /\.html$/,
                loader: 'html-loader?attrs=false'
            },
            {
                test: /\.(png|jpe?g|gif|svg|woff|woff2|ttf|eot|ico)$/,
                loader: 'null-loader'
            }
        ]
    },

    plugins: [
        new webpack.ContextReplacementPlugin(
            // The (\\|\/) piece accounts for path separators in *nix and Windows
            /(ionic-angular)|(angular(\\|\/)core(\\|\/)@angular)/,
            root('./src'), // location of your src
            {} // a map of your routes
        )
    ]
};

function environmentPath(env) {
    var filePath = './src/environments/environment' + (env === '' ? '' : '.' + env) + '.ts';
    if (fs.existsSync(filePath)) {
        return filePath;
    }

    console.log(chalk.red('\n' + filePath + ' does not exist!\n'));
    process.exit();
}

function root(localPath) {
    return path.resolve(__dirname, localPath);
}
```

**/package.json**

```json
....
        "ionic": "ionic",
        "build": "ionic build",
        "build:test": "ionic build --env=test",
        "build:prod": "ionic build --prod",
        "build:android": "ionic cordova build android --dev",
        "build:android:test": "ionic cordova build android --prod --env=test",
        "build:android:prod": "ionic cordova build android --prod",
....
```

**Comment by raajmayukh** [comment]
@keithdmoore Thanks for your idea to fix this solution. I have followed your solution and able to load files as per what's been set as fileName. But with this solution is it possible to access process.argv.env in TS files anywhere in the application?

I need to access them in my app to load some Mock services for different environments. Earlier(Angular 4 and ionic app scripts 1.x) i was achieving this by setting process.argv.NODE_ENV. I was using Opaque token for the same. Now with Angular 5 and ionic app scripts 3.1.x i am using Injection token and trying to access process.argv.env together with your solution but it always comes as undefined in TS files. Any idea on this?

I had created an issue earlier in Ionic repo, probably should have created here: Here is the link with all info:

 #[(https://github.com/ionic-team/ionic/issues/14150)]

**Comment by keithdmoore** [comment]
I’m sure you could. I followed someone else’s suggestion above.  I have a name attribute in my environment model that is the same as what is in the file name.  I can access the name in my ts files if I need to.

**Comment by tabirkeland** [comment]
All, with Ionic 4 release approaching, I know that ENV vars will be handled. If you are using Ionic 3.9.2, here is what I have been using for months to handle ENV vars. It's been pieced together from this and a few other threads.

https://gist.github.com/tabirkeland/a17c67b2f1ea3331d94db34ed7191c34

**Comment by writer0713** [comment]
made a starter project with detail explanations.

[https://github.com/writer0713/ionic-environment-setting/blob/master/README.md](https://github.com/writer0713/ionic-environment-setting/blob/master/README.md)

**Comment by tabirkeland** [comment]
@writer0713 nice one.  So I assume my gist helped?

**Comment by writer0713** [comment]
@tabirkeland 
Yes, I read so many different articles, but nothing worked. After I followed your code, it worked. Thanks.

**Comment by daraul** [comment]
@writer0713 your readme was a life saver. Worked like a charm
Thanks to you and @tabirkeland

**Comment by hsusyh** [comment]
> made a starter project with detail explanations.
> 
> https://github.com/writer0713/ionic-environment-setting/blob/master/README.md

@writer0713 @tabirkeland thanks!! It works so well!

**Comment by sana-moussa** [comment]
> I ended up using a combo of @emcniece and @gshigeto solutions, so massive thanks to them.
> 
> I wanted to have environment variables work in Ionic exactly like how they do in Angular CLI (especially given Ionic 4 is somewhere on the horizon), whilst also supporting Karma and Protractor tests.
> 
> To use the environment variables, just add `import { environment } from '@env/environment';`.
> Also supports `ngc` so I can build an Android APK with any environment config e.g. my `environment.test.ts` config.
> 
> This solution also supports using `ionic serve` with any of your configurations e.g.
> `ionic serve --env=prod`
> 
> I've managed all this with the following setup:
> 
> **/src/app/environments** folder contains
> 
> ```
> environment.model.ts
> environment.ts
> environment.prod.ts
> environment.test.ts
> ```
> **/tsconfig.json**
> 
> ```json
> {
>   "compilerOptions": {
>     "allowSyntheticDefaultImports": true,
>     "baseUrl": "./src",
>     "declaration": false,
>     "emitDecoratorMetadata": true,
>     "experimentalDecorators": true,
>     "paths": {   
>       "@env/*": [
>         "environments/*"
>       ]
>     },
>     "lib": [
>       "dom",
>       "es2016"
>     ],
>     "module": "es2015",
>     "moduleResolution": "node",
>     "sourceMap": true,
>     "target": "es5"
>   },
>   "include": [
>     "src/**/*.ts"
>   ],
>   "exclude": [
>     "node_modules",
>     "src/**/*.spec.ts"
>   ],
>   "typeRoots": [
>     "node_modules/@types"
>   ],
>   "compileOnSave": false,
>   "atom": {
>     "rewriteTsconfig": false
>   }
> }
> ```
> 
> **/config/webpack.config.js**
> 
> ```js
> var chalk = require("chalk");
> var defaultConfig = require('@ionic/app-scripts/config/webpack.config.js');
> var ionic_env = process.env.IONIC_ENV;
> var fs = require('fs');
> var path = require('path');
> var env = require('minimist')(process.argv.slice(2)).env || process.env.IONIC_ENV || 'dev';
> var webpack = require('webpack');
> 
> defaultConfig.dev.resolve.alias = {
>     "@env/environment": path.resolve(environmentPath('dev'))
> };
> 
> defaultConfig.prod.resolve.alias = {
>     "@env/environment": path.resolve(environmentPath('prod'))
> };
> 
> console.log(chalk.yellow.bgBlack('\nUsing ' + env + ' environment variables.\n'));
> 
> if (!!env) {
>     var pathToCustomEnvFile = path.resolve(environmentPath(env));
> 
>     defaultConfig[env] = defaultConfig[ionic_env];
>     defaultConfig[env].resolve.alias = {
>         "@env/environment": pathToCustomEnvFile
>     };
> 
>     var customEnvPlugin = new webpack.NormalModuleReplacementPlugin(
>         /src\/environments\/environment\.ts/, pathToCustomEnvFile
>     );
> 
>     defaultConfig.prod.plugins.push(customEnvPlugin);
>     defaultConfig.dev.plugins.push(customEnvPlugin);
> }
> 
> module.exports = function() {
>     return defaultConfig;
> };
> 
> 
> function environmentPath(env) {
>     var filePath = './src/environments/environment' + (env === '' ? '' : '.' + env) + '.ts';
>     if (fs.existsSync(filePath)) {
>         return filePath;
>     }
> 
>     console.log(chalk.red.bgWhite('\n' + filePath + ' does not exist!\n'));
>     process.exit();
> }
> ```
> 
> **/test-config/webpack.test.js**
> 
> ```js
> var webpack = require('webpack');
> var path = require('path');
> var fs = require('fs');
> var chalk = require('chalk');
> 
> module.exports = {
>     devtool: 'inline-source-map',
> 
>     resolve: {
>         alias: {
>             '@env/environment': path.resolve(environmentPath('dev')),
>         },
>         extensions: ['.ts', '.js']
>     },
> 
>     module: {
>         rules: [{
>                 test: /\.ts$/,
>                 loaders: [{
>                     loader: 'ts-loader'
>                 }, 'angular2-template-loader']
>             },
>             {
>                 test: /\.html$/,
>                 loader: 'html-loader?attrs=false'
>             },
>             {
>                 test: /\.(png|jpe?g|gif|svg|woff|woff2|ttf|eot|ico)$/,
>                 loader: 'null-loader'
>             }
>         ]
>     },
> 
>     plugins: [
>         new webpack.ContextReplacementPlugin(
>             // The (\\|\/) piece accounts for path separators in *nix and Windows
>             /(ionic-angular)|(angular(\\|\/)core(\\|\/)@angular)/,
>             root('./src'), // location of your src
>             {} // a map of your routes
>         )
>     ]
> };
> 
> function environmentPath(env) {
>     var filePath = './src/environments/environment' + (env === '' ? '' : '.' + env) + '.ts';
>     if (fs.existsSync(filePath)) {
>         return filePath;
>     }
> 
>     console.log(chalk.red('\n' + filePath + ' does not exist!\n'));
>     process.exit();
> }
> 
> function root(localPath) {
>     return path.resolve(__dirname, localPath);
> }
> ```
> 
> **/package.json**
> 
> ```json
> ....
>         "ionic": "ionic",
>         "build": "ionic build",
>         "build:test": "ionic build --env=test",
>         "build:prod": "ionic build --prod",
>         "build:android": "ionic cordova build android --dev",
>         "build:android:test": "ionic cordova build android --prod --env=test",
>         "build:android:prod": "ionic cordova build android --prod",
> ....
> ```

this is the only efficient solution that worked for me, thanks!

**Comment by elbidone** [comment]
@GFoley83 I've tested your solution but with AOT in specific cases it doesn't work, I ignore the reason.
How to reproduce the error: 
1. log any attribute from your env file on app.module and app.component. The attribute should have different value for each env.
3. run AOT build with --env=prod `ionic build --env=prod --prod`
2. you should observe on your console that app.module log value from dev and app.component log value from prod

It's annoying if you use firebase and need to initiate it with env vars on your app.module (`AngularFireModule.initializeApp(ENV.firebase)`). At the moment I didn't find a good solution. @writer0713 suggest to rewrite env file using npm hooks, I'll check it.
